### PR TITLE
token and nft contracts and unit tests

### PR DESCRIPTION
**Pull request by writetoyan**

### CODE REVIEW

**Review by fernandoandreotti — CHANGES_REQUESTED** [review]
There are many files here that we shouldnt commit (`artifacts/`, `cache/`, etc). Could you add a .gitignore like this please?
https://github.com/pcaversaccio/hardhat-project-template-ts/blob/main/.gitignore

File: backend/contracts/BetOnChain.sol [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    function mintBeginnerNft() external {
+        if (numberOfBets[msg.sender] < achievementRequirement.beginnerNumberOfBets) {
+            revert BetOnChain__RequirementsNotMet();
+        } 
+        bocNFT.safeMint(msg.sender, achievementURI.beginner);
+    }
+
+    function mintWarriorNft() external {
```
Comment (fernandoandreotti) [inline backend/contracts/BetOnChain.sol]: I agree there is a bit of duplication here. We could probably have one function to mint at any of the levels, provided some switch statement.

File: backend/contracts/BetOnChain.sol [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+pragma solidity 0.8.18;
+
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {BocNFT} from './BocNFT.sol';
+import {BocToken} from './BocToken.sol';
+
+error BetOnChain__RequirementsNotMet();
+
+contract BetOnChain is Ownable {
```
Comment (fernandoandreotti) [inline backend/contracts/BetOnChain.sol:12]: This is just a comment, we should do this on a separate ticket.
 
 
This is a good start, I think we can base it a lot more on the Lottery contract, with start/end times for bets, possibility of winner(s) to withdraw the available prizes. In addition it should be a contract factory, like in https://blog.logrocket.com/cloning-solidity-smart-contracts-factory-pattern/.

**Review by 0abst — APPROVED** [review]
I could run the tests, it works:

image

**Comment by 0abst** [comment]
Then just before the merge Conflicting files in:

 backend/hardhat.config.ts
backend/package.json 

could be found. I trust that for  backend/hardhat.config.ts the solution in main is better:

 
image

For backend/package.json the code of  "token and nft contracts and unit tests #2 " has been kept

image

**Pull request merged**